api/cluster/resource/templater.go

```diff
@@ -550,6 +596,120 @@ func createLoggerSpec(loggerURL string, loggerConfig models.LoggerConfig) *kserv
+		modelService.DeploymentMode == deployment.EmptyDeploymentMode {
+		var err error
+		newRevisionName, err = getNewRevisionNameForExistingSeverlessDeployment(
+			orig.Status.Components[component].LatestCreatedRevision,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("unable to generate new revision name: %w", err)
+		}
+	} else {
+		return nil, fmt.Errorf("invalid deployment mode: %s", modelService.DeploymentMode)
+	}
+	return appendPodSpreadingLabelSelectorsToTopologySpreadConstraints(
+		config.TopologySpreadConstraints,
+		newRevisionName,
+	)
+}
+
+// appendPodSpreadingLabelSelectorsToTopologySpreadConstraints makes a deep copy of the config topology spread
+// constraints and then adds the given revisionName as a label to the match labels of each topology spread constraint
+// to spread out all the pods across the specified topologyKey
+func appendPodSpreadingLabelSelectorsToTopologySpreadConstraints(
+	templateTopologySpreadConstraints []corev1.TopologySpreadConstraint,
+	revisionName string,
+) ([]corev1.TopologySpreadConstraint, error) {
+	topologySpreadConstraints, err := copyTopologySpreadConstraints(templateTopologySpreadConstraints)
+	if err != nil {
+		return nil, err
+	}
+	for i := range topologySpreadConstraints {
+		if topologySpreadConstraints[i].LabelSelector == nil {
+			topologySpreadConstraints[i].LabelSelector = &metav1.LabelSelector{
+				MatchLabels: map[string]string{"app": revisionName},
+			}
+		} else {
+			if topologySpreadConstraints[i].LabelSelector.MatchLabels == nil {
+				topologySpreadConstraints[i].LabelSelector.MatchLabels = make(map[string]string)
+			}
+			topologySpreadConstraints[i].LabelSelector.MatchLabels["app"] = revisionName
+		}
+	}
+	return topologySpreadConstraints, nil
+}
+
+// copyTopologySpreadConstraints copies the topology spread constraints using the service builder's as a template
+func copyTopologySpreadConstraints(
+	topologySpreadConstraints []corev1.TopologySpreadConstraint,
+) ([]corev1.TopologySpreadConstraint, error) {
+	topologySpreadConstraintsRaw, err := copystructure.Copy(topologySpreadConstraints)
+	if err != nil {
+		return nil, fmt.Errorf("Error copying topology spread constraints: %w", err)
+	}
+	topologySpreadConstraints, ok := topologySpreadConstraintsRaw.([]corev1.TopologySpreadConstraint)
+	if !ok {
+		return nil, fmt.Errorf("Error in type assertion of copied topology spread constraints interface: %w", err)
+	}
+	return topologySpreadConstraints, nil
+}
+
+// getNewRevisionNameForExistingSeverlessDeployment examines the current revision name of an inference service (
+// serverless deployment) app name that is given to it and increments the last value of the revision number by 1, e.g.
+// sklearn-sample-predictor-default-00001 -> sklearn-sample-predictor-default-00002
+func getNewRevisionNameForExistingSeverlessDeployment(currentRevisionName string) (string, error) {
+	revisionNameElements := strings.Split(currentRevisionName, "-")
```

Comment:
Should we handle the possibility of this split resulting in less than 2 parts (then, the `-1` in the next line would result in an out of bounds panic).

Reply:
Hmm in that case should we just perform a check that `len(revisionNameElements) < 4`? If it's less than 2 the name's probably going to be nonsensical anyway and we wouldn't know how to handle that. I chose 4 because the revision name's expected to be made up of at least 4 parts `[isvc name]-[predictor/transformer]-default-[6 digit 0-padded revision number]`. There may be more parts if the isvc name has additional hyphens.

Reply:
This probably isn't a perfect check either since someone might just happen to create a model with more hyphens than necessary in the `isvc name` , e.g. `sdk-test-1`, and the extra hyphens can count towards the length check (i.e. a revision name with `sdk-test-1-invalid-000001` would still pass the check).

Reply:
nit: can we add unit test for this function?

Reply:
Also, I'm not a huge fan of regex, but I think regex is more suitable here.